dll/win32/shlwapi/propbag.cpp

```diff
@@ -0,0 +1,254 @@
+    _Inout_ VARIANT *pvari,
+    _Inout_opt_ IErrorLog *pErrorLog)
+{
+    UNREFERENCED_PARAMETER(pErrorLog);
+
+    TRACE("%p: %s %p %p\n", this, debugstr_w(pszPropName), pvari, pErrorLog);
+
+    VARTYPE vt = V_VT(pvari);
+
+    ::VariantInit(pvari);
+
+#if (_WIN32_WINNT < _WIN32_WINNT_VISTA)
+    if ((m_dwMode & (STGM_READ | STGM_WRITE | STGM_READWRITE)) == STGM_WRITE)
+    {
+        ERR("%p: 0x%X\n", this, m_dwMode);
+        return E_ACCESSDENIED;
+    }
+#endif
+
+    if (!pszPropName || !pvari)
+    {
+        ERR("%p: %s %p %p\n", this, debugstr_w(pszPropName), pvari, pErrorLog);
+        return E_INVALIDARG;
+    }
+
+    INT iItem = m_PropMap.FindKey(pszPropName);
+    if (iItem == -1)
+    {
+        ERR("%p: %s %p %p\n", this, debugstr_w(pszPropName), pvari, pErrorLog);
+        return E_FAIL;
+    }
+
+    HRESULT hr = ::VariantCopy(pvari, &m_PropMap.GetValueAt(iItem));
+    if (FAILED(hr))
+    {
+        ERR("%p: 0x%08X %p\n", this, hr, pvari);
+        return hr;
+    }
+
+    hr = ::VariantChangeTypeForRead(pvari, vt);
+    if (FAILED(hr))
+    {
+        ERR("%p: 0x%08X %p\n", this, hr, pvari);
+        return hr;
+    }
+
+    return hr;
+}
+
+STDMETHODIMP
+CMemPropertyBag::Write(
+    _In_z_ LPCWSTR pszPropName,
+    _In_ VARIANT *pvari)
+{
+    TRACE("%p: %s %p\n", this, debugstr_w(pszPropName), pvari);
+
+#if (_WIN32_WINNT < _WIN32_WINNT_VISTA)
+    if ((m_dwMode & (STGM_READ | STGM_WRITE | STGM_READWRITE)) == STGM_READ)
+    {
+        ERR("%p: 0x%X\n", this, m_dwMode);
+        return E_ACCESSDENIED;
+    }
+#endif
+
+    if (!pszPropName || !pvari)
+    {
+        ERR("%p: %s %p\n", this, debugstr_w(pszPropName), pvari);
+        return E_INVALIDARG;
+    }
+
+    ATL::CComVariant vari;
+    HRESULT hr = vari.Copy(pvari);
+    if (FAILED(hr))
+    {
+        ERR("%p: %s %p: 0x%08X\n", this, debugstr_w(pszPropName), pvari, hr);
+        return hr;
+    }
+
+    if (!m_PropMap.SetAt(pszPropName, vari))
+    {
+        ERR("%p: %s %p\n", this, debugstr_w(pszPropName), pvari);
+        return E_FAIL;
+    }
+
+    return hr;
+}
+
+/**************************************************************************
+ *  SHCreatePropertyBagOnMemory (SHLWAPI.477)
+ *
+ * Creates a property bag object on memory.
+ *
+ * @param dwMode  Specifies either STGM_READ, STGM_WRITE or STGM_READWRITE. Ignored on Vista+.
+ * @param riid    Specifies either IID_IUnknown, IID_IPropertyBag or IID_IPropertyBag2.
+ *                Vista+ rejects IID_IPropertyBag2.
+ * @param ppvObj  Receives an IPropertyBag pointer.
+ * @return        An HRESULT value. S_OK on success, non-zero on failure.
+ * @see           http://undoc.airesoft.co.uk/shlwapi.dll/SHCreatePropertyBagOnMemory.php
+ */
+EXTERN_C HRESULT WINAPI
+SHCreatePropertyBagOnMemory(_In_ DWORD dwMode, _In_ REFIID riid, _Out_ void **ppvObj)
+{
+    TRACE("0x%08X, %s, %p\n", dwMode, debugstr_guid(&riid), ppvObj);
+
+    *ppvObj = NULL;
+
+    CMemPropertyBag* pMemBag = new CMemPropertyBag(dwMode);
```

Comment:
Dropping this directly into a `CComPtr` here will simplify cleanup because you don't have to call 'Release' anymore.

Reply:
Used `CComPtr`. Thanks.